File: auth/permissions.py
"""
Role -> permission matrix and helpers.

ADMIN receives every permission. USER receives a read-mostly subset plus the
sandbox simulation capability. Extend by editing ``_USER_PERMISSIONS`` or adding
new roles to ``ROLE_PERMISSIONS``.
"""

from __future__ import annotations

from typing import Dict, List, Set, Union

from auth.models import Permission, Role

# Permissions granted to the USER (analyst / viewer) role.
_USER_PERMISSIONS: Set[Permission] = {
    Permission.VIEW_LANDING,
    Permission.VIEW_DASHBOARD,
    Permission.VIEW_EXECUTIVE_SUMMARY,
    Permission.VIEW_FORECASTS,
    Permission.VIEW_INSIGHTS,
    Permission.VIEW_AGGREGATED_DATA,
    Permission.VIEW_MARKET_MONITOR,
    Permission.VIEW_AUDIT_SUMMARY,
    Permission.VIEW_WARRANTY,
    Permission.RUN_SANDBOX_SIMULATION,
}

# ADMIN gets everything.
_ADMIN_PERMISSIONS: Set[Permission] = set(Permission)

ROLE_PERMISSIONS: Dict[Role, Set[Permission]] = {
    Role.ADMIN: _ADMIN_PERMISSIONS,
    Role.USER: _USER_PERMISSIONS,
}


def _coerce_role(role: Union[Role, str]) -> Role:
    if isinstance(role, Role):
        return role
    return Role(role)


def _coerce_permission(permission: Union[Permission, str]) -> Permission:
    if isinstance(permission, Permission):
        return permission
    return Permission(permission)


def has_permission(role: Union[Role, str], permission: Union[Permission, str]) -> bool:
    """Return True if ``role`` is granted ``permission``."""
    try:
        r = _coerce_role(role)
        p = _coerce_permission(permission)
    except ValueError:
        return False
    return p in ROLE_PERMISSIONS.get(r, set())


def permissions_for_role(role: Union[Role, str]) -> List[str]:
    """Return the sorted list of permission string values for ``role``."""
    try:
        r = _coerce_role(role)
    except ValueError:
        return []
    return sorted(p.value for p in ROLE_PERMISSIONS.get(r, set()))


__all__ = ["ROLE_PERMISSIONS", "has_permission", "permissions_for_role"]

File: frontend/src/auth/permissions.js
// Permission strings mirroring the backend RBAC contract.
// USER (Analyst/Viewer) has a subset; ADMIN has all.
export const PERMISSIONS = {
  VIEW_DASHBOARD: 'view_dashboard',
  VIEW_EXECUTIVE_SUMMARY: 'view_executive_summary',
  VIEW_FORECASTS: 'view_forecasts',
  VIEW_INSIGHTS: 'view_insights',
  VIEW_AGGREGATED_DATA: 'view_aggregated_data',
  VIEW_MARKET_MONITOR: 'view_market_monitor',
  VIEW_AUDIT_SUMMARY: 'view_audit_summary',
  VIEW_WARRANTY: 'view_warranty',
  RUN_SANDBOX_SIMULATION: 'run_sandbox_simulation',
  RUN_SIMULATION: 'run_simulation',
  EDIT_SCENARIOS: 'edit_scenarios',
  VIEW_RAW_DATA: 'view_raw_data',
  MANAGE_THRESHOLDS: 'manage_thresholds',
  TRIGGER_RETRAINING: 'trigger_retraining',
  TRIGGER_DATA_FETCH: 'trigger_data_fetch',
  VIEW_AUDIT_FULL: 'view_audit_full',
  EXPORT_REPORTS: 'export_reports',
  REGENERATE_NARRATIVES: 'regenerate_narratives',
  MANAGE_USERS: 'manage_users',
};

/**
 * Returns true if the given user holds the given permission.
 */
export function can(user, perm) {
  if (!user || !perm) return false;
  return Array.isArray(user.permissions) && user.permissions.includes(perm);
}

/**
 * Returns true if the user has an admin role (or holds manage_users).
 */
export function isAdmin(user) {
  if (!user) return false;
  const role = (user.role || '').toUpperCase();
  return role === 'ADMIN' || can(user, PERMISSIONS.MANAGE_USERS);
}
